boost: use b2 from boost-build for building

sternenseemann opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

* Fixes a few cross compilation issues which broke e. g. `pkgsLLVM.boost`
* Gets rid of hacks to make boost understand nixpkgs' `CC_FOR_BUILD`
* Should overall decrease build time for boost a little bit

See commit messages for more details.

Built all `boost` versions on `x86_64-linux`, but darwin needs testing, I suspect we may need to reintroduce the removed hack for that in some way.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

sternenseemann committed:
boost-build: allow supplying a boost version to build b2 for

The `useBoost` argument expects an attribute set with an src and version
attribute (so a boost derivation works) and builds b2 for the given
version of boost. This is useful when bootstrapping boost: We can
override boost-build to get a boost-build derivation matching the
current boost version we want to build (b2 is not backwards compatible
enough to build older boost versions) without the need of having
one boost-build${version} attribute for every boost version we have
which is not very useful for nixpkgs' users.

sternenseemann committed:
boost*: use packaged b2 instead it of building in configurePhase

If we build two things in one derivation, it becomes more complicated
and its build time is extended. Therefore we should avoid this if
possible. There's a good opportunity for this with boost: We have
boost-build packaged already. This has the additional benefit that
we can get rid of $CC_FOR_BUILD entirely in boost, meaning we don't
need to rely on (as many) hacks to make boost understand our way of
cross compiling.

Unfortunately boost-build is not backwards compatible, so we need to
build a specific boost-build for each boost derivation (the number
could probably be reduced, but I'm not interested in testing a lot
of boost builds at the moment).

Additionally we fix a few cross compilation problems:

- boost couldn't cope with different types of compilers for native
  and cross (as happens if useLLVM is true). Since we only use one
  of them per derivation, this is no longer an issue.

- boost didn't find the cross ar and ranlib for compilation (since
  it doesn't check $AR or $RANLIB). Instead it used plain ar and
  ranlib form $PATH which were the native ones before. This is now
  fixed by setting these tools explicitly in user-config.jam (and
  no longer providing the native tools).

With these changes, pkgsLLVM.boost builds.

On darwin, instead of patching the clang-darwin.jam definition, we
instead supply -rpath $out/lib via <linkflags> which causes the
correct directory to be added to the libraries' rpaths, so that
they find each other.

sternenseemann commented:
@ofborg build boost boost-build boost155 boost159 boost15x boost160 boost165 boost166 boost167 boost168 boost169 boost16x boost170 boost171 boost172 boost173 boost174 boost175 boost17x

symphorien commented on pkgs/top-level/all-packages.nix:
```diff
@@ -14815,22 +14815,25 @@ in
 
-  boost155 = callPackage ../development/libraries/boost/1.55.nix { };
-  boost159 = callPackage ../development/libraries/boost/1.59.nix { };
+  inherit (callPackage ../development/libraries/boost { inherit (buildPackages) boost-build; })
```
```suggestion
  inherit (callPackages ../development/libraries/boost { inherit (buildPackages) boost-build; })
```

sternenseemann replied:
Thanks!

sternenseemann replied:
Actually had to revert that, since `callPackages` messes with the `override` of the inner attributes. It's fine how it is currently since we `callPackage` the derivations.

symphorien replied:
ah I see, that makes sense.

roberth replied:
It needs to mess with each package to do the splicing magic for cross compilation; the machinery that lets you use the same variable to represent multiple possible derivations, a different one of which is used in each of `buildInputs`, `nativeBuildInputs`, `deps${x}${y}`.

If you do this one level too soon, as you do here, the decision is made too soon, resulting in hard to understand cross breakage.

sternenseemann replied:
So how do I best put this in another file? `import`? Not sure what you are reffering to exactly, tbh.

roberth replied:
I probably did a terrible job trying to explain it. Thing is, it's underdocumented. The manual has this line

> To make this work, we “splice” together the six pkgsFooBar package sets and have callPackage actually take its arguments from that.

Which is hidden away in a cross-compilation section. This should also be part of the `callPackage` doc. Oh, which is never introduced.

I don't understand why it doesn't work for you. Maybe @Ericson2314 has experienced this?

`import` and many `callPackage` calls should work like it did before. That could be a way out.

sternenseemann replied:
It works for me at his, I thought you were implying using `callPackage` in the way I am would be a problem. So far I have seen no evidence of that, e. g. `pkgsLLVM.nix` builds fine which does rely on cross-compilation.

roberth replied:
Maybe it still works because boost is always used in `buildInputs`, which is the one attribute where a spliced package produces the same result as a non-spliced one.
It will break if it occurs as a dependency with a different [role](https://nixos.org/manual/nixpkgs/unstable/#variables-specifying-dependencies). This is extremely rare for boost, if not always wrong. I've found one weird example, `pkgs/applications/science/electronics/horizon-eda/default.nix` but I doubt its correctness. One example after looking way too hard, like 40% of usages. It might be ok to break.

sternenseemann replied:
Oh yeah, indeed. It does bug me a bit though, will try to find a way to solve this.

sternenseemann replied:
Nevermind, this wasn't even working before: In `nixos-unstable`:

```
nix-repl> callPackage ({ boost }: boost ? __spliced && boost ? nativeDrv && boost ? crossDrv) { }
false
```

Maybe something for a future refactor.

roberth replied:
These are only present when `pkgs` is cross.

```
nix-repl> pkgs = import ./. { crossSystem = "aarch64-linux"; localSystem = "x86_64-linux"; }          

nix-repl> pkgs.callPackage ({ boost }: boost?__spliced && boost?crossDrv && boost?nativeDrv) {}
true

```

So it does work.

sternenseemann replied:
Of course, yeah, can' think anymore today. On the plus side, it also seems to work with this PR applied, so that's good news!

roberth replied:
Isn't `inherit (buildPackages) boost-build;` effectively implied when you're only using it in `nativeBuildInputs`?

sternenseemann replied:
Yes, but since we `override` the `boost-build` derivation later, the splicing magic won't be able to process it and pass it through as is.

I think it is much nicer to use `buildPackages` explicitly here to work around it than to do `boost-build.nativeDrv.override`…

roberth replied:
I'm not convinced that splicing can't work with `override` yet, so I've split this off into #132651, which is in no way a blocker of course.

roberth replied:
I also noticed that the splicing logic traverses non-drv attrsets, so that explains why it still works.

sternenseemann replied:
I think it works because every derivation we inherit is a result of three nested `callPackage` calls (lol).

sternenseemann commented:
@ofborg build pkgsLLVM.nix

sternenseemann commented:
@ofborg build boost boost-build boost155 boost159 boost15x boost160 boost165 boost166 boost167 boost168 boost169 boost16x boost170 boost171 boost172 boost173 boost174 boost175 boost17x

sternenseemann commented:
@ofborg build python3Packages.boost

sternenseemann commented:
The two boost versions that are failing on darwin are already failing on master: [boost166](https://hydra.nixos.org/job/nixpkgs/trunk/boost166.x86_64-darwin) and [boost168](https://hydra.nixos.org/job/nixpkgs/trunk/boost168.x86_64-darwin).

In that sense, I think this is good to go in principle: All versions build for aarch64-linux, x86_64-linux and no new regressions for darwin. Will rebase on staging now.

sternenseemann commented:
Nevermind, boost on darwin does build, but downstream users are broken. Will need to investigate how to best fix this.

SuperSandro2000 commented on pkgs/development/tools/boost-build/default.nix:
```diff
@@ -32,8 +50,13 @@ stdenv.mkDerivation rec {
+
+    # older versions of b2 created this symlink,
+    # which we want to support building via useBoost.
+    test -e "$out/bin/bjam" || ln -s b2 "$out/bin/bjam"
```
```suggestion
    [ ! -e "$out/bin/bjam" ] && ln -s b2 "$out/bin/bjam"
```

sternenseemann replied:
Reasoning?

SuperSandro2000 replied:
test is uncommon to be used and this makes the condition clearer.

roberth replied:
I prefer the `||` though.

I also use `test` which fits more natural into the language. The `]` is nonsense.

SuperSandro2000 replied:
whatever, UNRELATED

SuperSandro2000 commented on pkgs/development/libraries/boost/generic.nix:
```diff
@@ -173,13 +179,13 @@ stdenv.mkDerivation {
   ] ++ optional enablePython "--with-python=${python.interpreter}"
-    ++ [ (if stdenv.hostPlatform == stdenv.buildPlatform then "--with-icu=${icu.dev}" else "--without-icu") ]
-    ++ optional (toolset != null) "--with-toolset=${toolset}";
+    ++ [ (if stdenv.hostPlatform == stdenv.buildPlatform then "--with-icu=${icu.dev}" else "--without-icu") ];
```
```suggestion
    ++ [ (if stdenv.hostPlatform == stdenv.buildPlatform then "--with-icu=${lib.getDev icu}" else "--without-icu") ];
```

sternenseemann replied:
Unrelated change.

SuperSandro2000 commented on pkgs/development/libraries/boost/generic.nix:
```diff
@@ -1,7 +1,7 @@
 , fetchpatch
 , which
-, buildPackages
 , toolset ? /**/ if stdenv.cc.isClang  then "clang"
             else null
```
```suggestion
, toolset ? if stdenv.cc.isClang  then "clang" else null
```

sternenseemann replied:
Unrelated change.

sternenseemann commented:
@ofborg build nix

Fixed the issue on darwin now, `nix` builds now which was previously failing with unresolved symbols in `checkPhase`. This was because boost's build doesn't actually set an `rpath` for the libraries it produces, so `dyld` can't resolve the references to `@rpath` when the boost libs load each other.

sternenseemann commented:
@ofborg build boost boost-build boost155 boost159 boost15x boost160 boost165 boost166 boost167 boost168 boost169 boost16x boost170 boost171 boost172 boost173 boost174 boost175 boost17x

oxalica commented:
It failed to cross to riscv64 by `nix build -f . pkgsCross.riscv64.boost175` while it can be built without this PR.
[Build log](https://gist.github.com/oxalica/0f35d9619d2aad5f46f9b851c2a16f7e)
The log shows it is built successfully but unexpectedly fail just before installing.

sternenseemann commented:
Huh, usually this means that there's a failure somewhere before that (which are really hard to spot) which causes the build as a whole to fail (quite late). The only thing I could find is the following:

```
/nix/store/xpqia918b3l2gx7sj2b2is04a49v88c8-riscv64-unknown-linux-gnu-binutils-2.35.1/bin/riscv64-unknown-linux-gnu-ld: bin.v2/libs/test/build/gcc-cross/release/threading-multi/visibility-hidden/execution_monitor.o: in function `.L0 ':
execution_monitor.cpp:(.text+0xa92): warning: fedisableexcept is not implemented and will always fail
/nix/store/xpqia918b3l2gx7sj2b2is04a49v88c8-riscv64-unknown-linux-gnu-binutils-2.35.1/bin/riscv64-unknown-linux-gnu-ld: execution_monitor.cpp:(.text+0xa62): warning: feenableexcept is not implemented and will always fail
```

Not sure how serious this is, but “will always fail” sounds like it maybe could be the culprit?

sternenseemann committed:
boost: always set toolset for known compilers

There is an edge case when cross compiling where the auto detection
script would not correctly recognize the compiler (as it is only
good at recognizing native compilers, really, which we don't have
anymore since we don't need the build->build one anymore).

If bootstrap.sh doesn't detect the compiler correctly, it'll generate
a project-config.jam with a syntax error which breaks the build in a
hard to spot way: only a warning is displayed after configuring, the
build will appear to run normally until it fails quietly just before
installing. By providing it explicitly, we can prevent this.

sternenseemann commented:
@ofborg build build boost boost-build boost155 boost159 boost15x boost160 boost165 boost166 boost167 boost168 boost169 boost16x boost170 boost171 boost172 boost173 boost174 boost175 boost17x nix pkgsCross.riscv64.boost175 pkgsCross.aarch64-multiplatform.boost

@oxalica should be fixed in 43e0f0688a19bbd88cdf898efb7b14ecf09717c4

sternenseemann commented:
Let's merge this? Any concerns left?
